packages/elements/src/react/common/form/otp.tsx

```diff
@@ -45,120 +83,314 @@ export const OTPInput = React.forwardRef<HTMLInputElement, OTPInputProps>(functi
       }
-      `}</style>
+        `}
+        </style>
         <RadixControl
           ref={innerRef}
           {...rest}
-          onBlur={event => {
-            setSelectionRange([-1, -1]);
-            rest?.onBlur?.(event);
+          value={value}
+          onSelect={e => {
+            _select();
+            rest?.onSelect?.(e);
+          }}
+          onTouchEnd={e => {
+            const isFocusing = document.activeElement === e.currentTarget;
+            if (isFocusing) {
+              _select();
+            }
+
+            rest?.onTouchEnd?.(e);
+          }}
+          onTouchMove={e => {
+            const isFocusing = document.activeElement === e.currentTarget;
+            if (isFocusing) {
+              _select();
+            }
+
+            rest?.onTouchMove?.(e);
+          }}
+          onDoubleClick={e => {
+            const isFocusing = document.activeElement === e.currentTarget;
+
+            if (isFocusing) {
+              e.currentTarget.setSelectionRange(0, e.currentTarget.value.length);
+              _select();
+            }
+
+            rest?.onDoubleClick?.(e);
+          }}
+          onInput={e => {
+            _select();
+
+            rest?.onInput?.(e);
           }}
-          onSelect={event => {
-            setSelectionRange(cur => selectionRangeUpdater(cur, innerRef));
-            rest?.onSelect?.(event);
+          onKeyDown={e => {
+            _keyDown(e);
+            _select();
+
+            rest?.onKeyDown?.(e);
+          }}
+          onKeyUp={e => {
+            _select();
+
+            rest?.onKeyUp?.(e);
+          }}
+          onFocus={e => {
+            if (innerRef.current) {
+              const start = Math.min(innerRef.current.value.length, length - 1);
+              const end = innerRef.current.value.length;
+              innerRef.current.setSelectionRange(start, end);
+              setSelectionRange([start, end]);
+            }
+
+            rest?.onFocus?.(e);
           }}
           style={inputStyle}
+          data-otp-input-segmented
         />
         <div
-          className={className}
+          className={userProvidedClassName}
           aria-hidden
-          style={{
-            zIndex: 1,
-            pointerEvents: 'none',
-          }}
+          style={renderStyle}
+          data-otp-input-segmented-render
         >
-          {Array.from({ length }).map((_, i) => (
-            <React.Fragment key={`otp-segment-${i}`}>
-              {render({
-                value: String(props.value)[i] || '',
-                status:
-                  selectionRange[0] === selectionRange[1] && selectionRange[0] === i
-                    ? 'cursor'
-                    : selectionRange[0] <= i && selectionRange[1] > i
-                    ? 'selected'
-                    : 'none',
-                index: i,
-              })}
-            </React.Fragment>
-          ))}
+          {Array.from({ length: length }).map((_, i) => {
```

Comment:
Cleaned up the inline ternary here and added `isFocused` to the logic. This replaces the previous `onBlur` handler that reset the selection when the input was not focused anymore.